packages/mdc-data-table/foundation.ts

```diff
@@ -0,0 +1,193 @@
+    for (let rowIndex = 0; rowIndex < this.adapter_.getRowCount(); rowIndex++) {
+      const rowId = this.adapter_.getRowIdAtIndex(rowIndex);
+
+      let isSelected = false;
+      if (rowId && rowIds.indexOf(rowId) >= 0) {
+        isSelected = true;
+      }
+
+      this.adapter_.setRowCheckboxCheckedAtIndex(rowIndex, isSelected);
+      this.selectRowAtIndex_(rowIndex, isSelected);
+    }
+
+    this.setHeaderRowCheckboxState_();
+  }
+
+  /**
+   * @return Returns array of selected row ids.
+   */
+  getSelectedRowIds(): Array<string|null> {
+    const selectedRowIds: Array<string|null> = [];
+    for (let rowIndex = 0; rowIndex < this.adapter_.getRowCount(); rowIndex++) {
+      if (this.adapter_.isCheckboxAtRowIndexChecked(rowIndex)) {
+        selectedRowIds.push(this.adapter_.getRowIdAtIndex(rowIndex));
+      }
+    }
+
+    return selectedRowIds;
+  }
+
+  /**
+   * Handles header row checkbox change event.
+   */
+  handleHeaderRowCheckboxChange() {
+    const isHeaderChecked = this.adapter_.isHeaderRowCheckboxChecked();
+
+    for (let rowIndex = 0; rowIndex < this.adapter_.getRowCount(); rowIndex++) {
+      this.adapter_.setRowCheckboxCheckedAtIndex(rowIndex, isHeaderChecked);
+      this.selectRowAtIndex_(rowIndex, isHeaderChecked);
+    }
+
+    if (isHeaderChecked) {
+      this.adapter_.notifySelectedAll();
+    } else {
+      this.adapter_.notifyUnselectedAll();
+    }
+  }
+
+  /**
+   * Handles change event originated from row checkboxes.
+   */
+  handleRowCheckboxChange(event: Event) {
+    const rowIndex = this.adapter_.getRowIndexByChildElement(event.target as HTMLInputElement);
+
+    if (rowIndex === -1) {
+      return;
+    }
+
+    const selected = this.adapter_.isCheckboxAtRowIndexChecked(rowIndex);
+
+    this.selectRowAtIndex_(rowIndex, selected);
+    this.setHeaderRowCheckboxState_();
+
+    const rowId = this.adapter_.getRowIdAtIndex(rowIndex);
+    this.adapter_.notifyRowSelectionChanged({rowId, rowIndex, selected});
+  }
+
+  /**
+   * Updates header row checkbox state based on number of rows selected.
+   */
+  private setHeaderRowCheckboxState_() {
+    if (this.adapter_.getSelectedRowCount() === this.adapter_.getRowCount()) {
+      this.adapter_.setHeaderRowCheckboxIndeterminate(false);
+      this.adapter_.setHeaderRowCheckboxChecked(true);
+    } else if (this.adapter_.getSelectedRowCount() === 0) {
+      this.adapter_.setHeaderRowCheckboxIndeterminate(false);
+      this.adapter_.setHeaderRowCheckboxChecked(false);
+    } else {
+      this.adapter_.setHeaderRowCheckboxChecked(false);
+      this.adapter_.setHeaderRowCheckboxIndeterminate(true);
+    }
+  }
+
+  /**
+   * Sets the attributes of row element based on selection state.
+   */
+  private selectRowAtIndex_(rowIndex: number, selected: boolean) {
+    if (selected) {
+      this.adapter_.addClassAtRowIndex(rowIndex, cssClasses.ROW_SELECTED);
+      this.adapter_.setAttributeAtRowIndex(rowIndex, strings.ARIA_SELECTED, 'true');
```

Comment:
could this just be:
```
if (selected) {
      this.adapter_.addClassAtRowIndex(rowIndex, cssClasses.ROW_SELECTED);
} else {
      this.adapter_.removeClassAtRowIndex(rowIndex, cssClasses.ROW_SELECTED);
}
this.adapter_.setAttributeAtRowIndex(rowIndex, strings.ARIA_SELECTED, selected);
```

Reply:
setAttribute accepts string value for 'true' / 'false', we would still need ternary operator if we do that. :)

Reply:
ah right right...for some reason i thought it casts it for free